runtime/parachains/src/paras_inherent/mod.rs

```diff
@@ -277,18 +283,28 @@ impl<T: Config> Pallet<T> {
 		);
 
 		let now = <frame_system::Pallet<T>>::block_number();
+		let expected_bits = <scheduler::Pallet<T>>::availability_cores().len();
 
 		let mut candidate_weight = backed_candidates_weight::<T>(&backed_candidates);
 		let mut bitfields_weight = signed_bitfields_weight::<T>(signed_bitfields.len());
 		let disputes_weight = dispute_statements_weight::<T>(&disputes);
+		let mut max_enact_candidates_weight = enact_candidates_weight::<T>(
+			bitfields_count_ones(
+				&signed_bitfields,
+				expected_bits,
```

Comment:
It feels odd that the check for bitfield lengths moved in here

Reply:
I would not say its "moved" in here, but instead we do the check here in addition to the places we already did the check. In other words we abstract out some checks to `cheap_bitfield_checks` and do them in `bitfield_count_ones` and then later on in `process_bitfields`. 

Now that I think about it though, we could do some "lightweight" sanitization upfront where we just return the bitfields that pass `cheap_bitfield_checks`

Reply:
Actually realizing it would open up DoS vector to do any sanitization prior to weight checks. The reason I consider `bitfields_count_ones` iteration relatively safe to a DoS attack is that we have a hard cap on the number of bitfields with iterate with `MAX_UNCHECKED_BITFIELD_ITERATIONS`:

https://github.com/paritytech/polkadot/blob/9efb72f826ce12721d8575a2e15cb7c7ade131ce/runtime/parachains/src/paras_inherent/weights.rs#L89-L92

Reply:
I made sure to add a note to `bitfield_count_ones` explaining this limitation https://github.com/paritytech/polkadot/blob/9efb72f826ce12721d8575a2e15cb7c7ade131ce/runtime/parachains/src/paras_inherent/weights.rs#L81-L82

In theory a malicious validator could craft a block where the `bitfields` is padded with `MAX_UNCHECKED_BITFIELD_ITERATIONS` count of bogus bitfields to skew the weight calc. But that seems like an ok trade off because while it will skew weights it would prevent some insane unbounded iteration

Reply:
This makes sense, we should also have an upper bound on the node side.

Reply:
https://github.com/paritytech/polkadot/issues/4512